edge-hub/test/Microsoft.Azure.Devices.Edge.Hub.E2E.Test/TestModule.cs

```diff
@@ -36,39 +37,52 @@ public static async Task<TestModule> CreateAndConnect(RegistryManager rm, string
 
         public Task SetupReceiveMessageHandler()
         {
-            this.received = new HashSet<int>();
-            return this.moduleClient.SetMessageHandlerAsync(this.MessageHandler, null);
+            this.receivedForInput["_"] = new HashSet<int>();
+            return this.moduleClient.SetMessageHandlerAsync(this.MessageHandler, this.receivedForInput["_"]);
+        }
+
+        public Task SetupReceiveMessageHandler(string input)
+        {
+            this.receivedForInput[input] = new HashSet<int>();
+            return this.moduleClient.SetInputMessageHandlerAsync(input, this.MessageHandler, this.receivedForInput[input]);
         }
 
         Task<MessageResponse> MessageHandler(Message message, object userContext)
         {
             int messageIndex = int.Parse(message.Properties["testId"]);
-            this.received.Add(messageIndex);
+            var received = userContext as ISet<int>;
+            received?.Add(messageIndex);
             return Task.FromResult(MessageResponse.Completed);
         }
 
-        public ISet<int> GetReceivedMessageIndices() => this.received;
+        public ISet<int> GetReceivedMessageIndices() => this.receivedForInput["_"];
+
+        public ISet<int> GetReceivedMessageIndices(string input) => this.receivedForInput[input];
+
+        public Task<int> SendMessagesByCountAsync(string output, int startIndex, int count, TimeSpan timeout) =>
+            this.SendMessagesByCountAsync(output, startIndex, count, timeout, TimeSpan.Zero);
 
-        public async Task<int> SendMessagesByCountAsync(string output, int startIndex, int count, TimeSpan timeout)
+        public async Task<int> SendMessagesByCountAsync(string output, int startIndex, int count, TimeSpan timeout, TimeSpan sleepTime)
         {
-            int sentMessagesCount = await this.SendMessagesAsync(output, startIndex, count, timeout);
+            int sentMessagesCount = await this.SendMessagesAsync(output, startIndex, count, timeout, sleepTime);
             if (sentMessagesCount < count)
             {
                 throw new TimeoutException($"Attempted to send {count} messages in {timeout.TotalSeconds} seconds, but was able to send only {sentMessagesCount}");
             }
             return sentMessagesCount;
         }
 
-        public Task<int> SendMessagesForDurationAsync(string output, TimeSpan duration) => this.SendMessagesAsync(output, 0, int.MaxValue, duration);
+        public Task<int> SendMessagesForDurationAsync(string output, TimeSpan duration) => this.SendMessagesAsync(output, 0, int.MaxValue, duration, TimeSpan.Zero);
 
-        async Task<int> SendMessagesAsync(string output, int startIndex, int count, TimeSpan duration)
+        async Task<int> SendMessagesAsync(string output, int startIndex, int count, TimeSpan duration, TimeSpan sleepTime)
```

Comment:
>int count, TimeSpan duration, TimeSpan sleepTime [](start = 73, length = 48)

Should you check that count*sleeptime <= duration,  print a warning so that the tester knows they might not receive all messages?